src/spaces/openstreetmap.jl

```diff
@@ -0,0 +1,402 @@
+"""
+function osm_plan_route(start::Int, finish::Int, model::ABM{A,OpenStreetMapSpace}) where {A}
+    #TODO: Expand to allow 'fastest_route' as well
+    route =
+        shortest_route(model.space.m, model.space.m.n[start], model.space.m.n[finish])[1]
+    map(p -> getindex(model.space.m.v, p), route)
+end
+
+"""
+    osm_map_coordinates(agent, model::ABM{A,OpenStreetMapSpace})
+
+Return a set of coordinates for an agent on the underlying map. Useful for plotting.
+"""
+function osm_map_coordinates(agent, model)
+    if agent.pos[1] != agent.pos[2]
+        start = get_ENU(agent.pos[1], model)
+        finish = get_ENU(agent.pos[2], model)
+        travelled = agent.pos[3] / Agents.osm_road_length(agent.pos, model)
+        (
+            getX(start) * (1 - travelled) + getX(finish) * travelled,
+            getY(start) * (1 - travelled) + getY(finish) * travelled,
+        )
+    else
+        position = get_ENU(agent.pos[1], model)
+        (getX(position), getY(position))
+    end
+end
+
+"""
+    osm_road_length(pos::OSMPos, model)
+    osm_road_length(start::Int, finish::Int, model)
+
+Returns the distance travelled between two intersections.
+"""
+osm_road_length(pos::OSMPos, model::ABM{<:AbstractAgent,<:OpenStreetMapSpace}) =
+    osm_road_length(pos[1], pos[2], model)
+osm_road_length(p1::Int, p2::Int, model::ABM{<:AbstractAgent,<:OpenStreetMapSpace}) =
+    model.space.m.w[p1, p2]
+
+#HELPERS, NOT EXPORTED
+
+get_ENU(pos::Int, model) = model.space.m.nodes[model.space.m.n[pos]]
+
+#######################################################################################
+# Agents.jl space API
+#######################################################################################
+
+random_position(model::ABM{<:AbstractAgent,<:OpenStreetMapSpace}) =
+    OSMPos(rand(1:nv(model)))
+
+function add_agent_to_space!(
+    agent::A,
+    model::ABM{A,<:OpenStreetMapSpace},
+) where {A<:AbstractAgent}
+    push!(model.space.s[agent.pos[1]], agent.id)
+    return agent
+end
+
+function remove_agent_from_space!(
+    agent::A,
+    model::ABM{A,<:OpenStreetMapSpace},
+) where {A<:AbstractAgent}
+    prev = model.space.s[agent.pos[1]]
+    ai = findfirst(i -> i == agent.id, prev)
+    deleteat!(prev, ai)
+    return agent
+end
+
+function move_agent!(agent::A, pos::OSMPos, model::ABM{A,<:OpenStreetMapSpace}) where {A}
+    remove_agent_from_space!(agent, model)
+    agent.pos = pos
+    add_agent_to_space!(agent, model)
+end
+
+"""
+    move_agent!(agent::A, model::ABM{A, OpenStreetMapSpace}, distance::Real)
+
+`distance` travelled in meters along an agents current route.
+"""
+function move_agent!(
+    agent::A,
+    model::ABM{A,<:OpenStreetMapSpace},
+    distance::Real,
+) where {A<:AbstractAgent}
+
+    #TODO: Assumption that an agent can only end its route on an intersection.
+    #It cannot pull up to someone's house, or park on the side of the road at present.
```

Comment:
I'm a bit confused by this comment. If the position of the agent is `(0, 1, 0.2)` and the road length between 0 and 1 is 4.0, then `move_agent!(..., 2.0)` will make the agent have position `(0, 1, 2.2)`, right? And therefore the movement indeed stopped in-between the nodes.

Reply:
Been a while since I wrote this, so I'm a bit rusty on the reason. What I recall is that the issue is not in `move_agent!` but in the route planner. As I said above, OSMX.jl only works with intersections as far as I can tell. So the call to `shortest_route` is between nodes not some way down an edge.

 In the example we use `finish = random_position(model) # At an intersection`, if we were to use `osm_random_direction` without extending the route planner I think we get into the situation where the route is finished, but there is still distance to travel and our stopping condition isn't met, but there's no understanding of where to move...

Reply:
We should instead have a `random_route` function instead then that optionally gets an input starting node.

Reply:
Not a bad idea. Will keep that an open possibility.

Reply:
Right, so the reason for this with the current implementation: `pos::Tuple{Int,Int,Float64}` gives us a location somewhere on a road or at an intersection&mdash;fine. OSMX.jl's route planner only uses intersections. It's not possible to plan a route to a specific location on a road. So we make `route::Vector{Int}`, since we can just take a stroll down each of the intersections until the end. To be able to allow an agent to finish somewhere other than an intersection, we need an additional `destination::Tuple{Int,Int,Float64}` property or perhaps something like `route::Tuple{Vec{Int},Tuple{Int,Int,Float64}}` which implies your destination is at `route[2]`, if you follow `route[1]...`. I'll think about the ramifications of this and take a stab at a better design.

Reply:
Is it really so important to make routes that do _not_ end at intersections? Given the complexity that it brings, I'm not sure whether it is _that_ beneficial. Such minute details may not matter for the actual scientific use of an ABM on a map, like e.g. spreading rate or other things, which anyways are calculated over several ABM realizations.

Reply:
It's not really added complexity, just an additional tuple. I'd argue it's definitely worth it if we're actually making this a space rather that an Integration Example.

If I want to measure traffic flows from my franchise of petrol stations that are sitting on highways (i.e. not on intersections) coming in and out of a city, I certainly don't want to have my estimations off by kilometres.

Reply:
Okay, you have a fair point I'm convinced.